### DISCUSSION TITLE
MUC Join/Leave Room

### DISCUSSION THREAD

**Discussion by OlHall** [vitalyster/SharpXMPP#144] [Q&A] [answered]
Hi,
Is there an example showing the correct procedure for joining/leaving a group chat room?
Thanks

**Comment by vitalyster** [comment, marked as answer]
This library automatically joins to "bookmarked" chat rooms, take a look at the [BookmarksManager code](https://github.com/vitalyster/SharpXMPP/blob/master/SharpXMPP.Shared/XMPP/Client/MUC/Bookmarks/BookmarksManager.cs#L55) if you want to join room by hand

**Reply by OlHall** [reply]
Yep, I tried following the code for joining a bookmarked room and replicate it for a "known" room JID, but it didn't work.

I saw your other replies about not seeing the need to provide documentation.  I'm afraid I think you underestimate how "readable" your code is.  If you make it a hassle to understand the structure of the code or how to use it, I would expect many people will simply move on to something else - which is a shame when you've done so much good work.

**Reply by vitalyster** [reply]
> it didn't work.

https://www.chiark.greenend.org.uk/~sgtatham/bugs.html

**Reply by OlHall** [reply]
Ok; if you want to be like that.

- I have created a test project for eval of your library SharpXMPP (.Net Framework 4.8)
- I have installed your latest Nuget
- I have written a simple test that connects to a development Openfire server
- I had to disable STARTTLS on the OpenFire server because I don't have a valid certificate for dev, and your library doesn't appear to offer a mechanism, to disable TLS.  This may be incorrect, but in the absence of documentation, this was my conclusion.
- The Openfire server has 2 group chat rooms configured, with permissions set for public access, no passwords etc.
- I have tested the rooms using the Spark XMPP client and I am able to join the rooms and chat
- I have also tested the Openfire configuration using MatriX vNext which successfully joins a room and sends a group chat message.
- I successfully connect to the Openfire server using SharpXMPP and can send direct messages.
- I then attempted to join a MUC room, and send/receive a group message.

My client SignedIn event fires correctly.

I then tried to join a group room manually by taking the BookmarkManager (which I'd found before you pointed me towards it) code as a template.
      
```
private void Join(JID roomJid)
        {
            var mucPresence = new XMPPPresence(_connection.Capabilities)
            {
                To = roomJid // new JID(string.Format("{0}/{1}", room.JID.BareJid, room.Nick))
            };
            var x = new XElement(XNamespace.Get(Namespaces.MUC) + "x");

            mucPresence.Add(x);
            _connection.Send(mucPresence);
        }
```

This code completes without any error, but also without any noticeable change in state.  Sending a client message does not deliver a message to the group room and group room messages do not trigger the client Message handler.

I then started this topic on Github asking for an example to assist me in understanding how the library should be used, to ensure I wasn't doing anything stupid, or missing something important.

You pointed me to the code that I had already tried to follow without a successful outcome.  I've since tried to use the BookmarkManager code as-is to establish where I am going wrong.  I have added this to my test code.

```
            BookmarksManager bm = new BookmarksManager(client);
            bm.BookmarksSynced += (XmppConnection sender) =>
            {
                foreach (BookmarkedConference room in bm.Rooms)
                {
                    Console.WriteLine("Room: " + room.JID.BareJid);
                }
            };
```

The client SignedIn event fires, the BoomarkManager SignedIn event handler fires.
The BookmarksSynced event handler is invoked for bookmarked rooms, which is understandable for bookmarked rooms, but not helpful if you just want to query the current rooms available.

I then call BoommarkManager.Join(...) on a room from the "BookmarkedConference" collection in the BookmarkManager, and nothing happens.  Neither Openfire, nor the Spark client (for obvious reasons) show the client joining the room, and sending/receiving messages also does not touch the room.

There's nothing in the code or documentation to hint at why.

Moving on...

I've now run the code directly rather than using the package, and I see why the room join is failing - the nick is null.  If I modify the nick in the Join XMPP presence message, it *does* now connect, bookmarked or otherwise.  So definitely progress.

I now receive messages from the room, but I don't seem to be able to send them TO the room from SharpXMPP.  I'm sure I'll work it out eventually.  type != groupchat?

---

I was hoping this library would remove the need to understand the internals of XMPP, but I already now know far more than I ever intended to.  It's like learning a language by reading the compiler source code...

**Reply by vitalyster** [reply]
> To = roomJid // new JID(string.Format("{0}/{1}", room.JID.BareJid, room.Nick))

So, it was not work because you are commented out correct code which sends nick? Great to know nothing to fix in the library

> remove the need to understand the internals of XMPP

You said you read previous discussions, take a look at https://github.com/vitalyster/SharpXMPP/discussions/121#discussioncomment-1384237

**Reply by OlHall** [reply]
Oof!

Yes, I did read comment 121.  I didn't think I was trying to _extend_ the library, just _use_ it.  And since that comment is from 2019, I don't feel too bad thinking things might have progressed.

And if you look at the first snippet, you'll see I'm passing **in** a complete JID.  The comment was just a reminder to what your library does.  If you read my full transcript, I did try this with your library code too, and that didn't work either.

The problem appears to be that the BookmarkManager doesn't (always?) add the nick to the rooms enumeration, so when you call Join, it tries to append ```null``` to the JID.

I fully admit I don't know a great deal about XMPP which is why I was looking at your library - so, if I may ask a question?

- When you join a group chat (MUC) room, the JID is comprised of the roomname@service.domain and the nick of the **client** is appended to the JID - i.e. roomname@service.domain/clientNick.  So should the BookmarkManager.Join method not be appending the client's nick, rather than something from the room (which in my case was null) or, if that is a property of a bookmarked room, at least test for null and default it to the current connection's nick?

I do feel as though you've been really defensive about all of this - I was only asking for some help/guidance.  If I've annoyed you, I apologise, and I'll move on to use a different implementation.
